### PR TITLE
Add tests for `sf::String`

### PR DESCRIPTION
**Pull request by ChrisThrasher**
## Description

These tests don't bother covering all edge or corner case behavior. Their intent is to test all the simple cases of `sf::String` and lay the groundwork for more sophisticated tests which may come later. I'm not experienced with wide strings or UTF-32 strings so I'm not even sure how to tackle testing the more complicated uses.

### CODE REVIEW

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Review by Bromeon — COMMENTED** [review]
Great work, thanks a lot!

I think the tests may benefit from using some actual wide-char or UTF-32 characters where applicable, so we don't just have ASCII. See https://stackoverflow.com/q/6796157 for string literals 🙂

File: test/System/String.test.cpp [hunk 1]
```diff
@@ -1,8 +1,370 @@
 #include <SFML/System/String.hpp>
 
+#include <doctest/doctest.h>
+
+#include <GraphicsUtil.hpp>
+#include <array>
 #include <type_traits>
```
Comment (eXpl0it3r) [inline test/System/String.test.cpp:7]: Shouldn't you technically also include `<string>`

File: test/System/String.test.cpp [hunk 2]
```diff
@@ -1,8 +1,370 @@
...
+        CHECK(string == "please don't insert text");
+    }
+
+    SUBCASE("find()")
+    {
+        const sf::String string("a little bit of this and a little bit of that");
+        CHECK(string.find("a little bit") == 0);
+        CHECK(string.find("a little bit", 15) == 25);
+        CHECK(string.find("a little bit", 1'000) == sf::String::InvalidPos);
+        CHECK(string.find("no way you find this") == sf::String::InvalidPos);
```
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/String.test.cpp:263]: I'm not sure what to make of that. Seems like perfectly valid C++. Those runners are using GCC 11 so it's certainly not an issue with a compiler having an incomplete implementation of C++17. @vittorioromeo Do you know why MinGW may be struggling with `sf::String::InvalidPos`?

[the remark this answers is not in the document]
Comment (binary1248) [reply test/System/String.test.cpp:263]: It's some tricky behaviour that got changed between C++11, C++14 and C++17.

https://stackoverflow.com/questions/8016780/undefined-reference-to-static-constexpr-char

Maybe if you add a definition in the .cpp the linker will have a symbol to export in the .dll.

[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/String.test.cpp:263]: I thought the whole point of C++17's `inline` variables want to prevent needing that extra definition?

[the remark this answers is not in the document]
Comment (binary1248) [reply test/System/String.test.cpp:263]: Compiler bugs exist. 😉

**Comment by ChrisThrasher** [comment]
I wonder why there's no code coverage comment on this PR

File: test/System/String.test.cpp [hunk 3]
```diff
@@ -1,8 +1,370 @@
...
+            CHECK(string.toUtf16() == std::basic_string<std::uint16_t>({'a'}));
+            CHECK(string.toUtf32() == std::basic_string<std::uint32_t>({'a'}));
+            CHECK(string.getSize() == 1);
+            CHECK(!string.isEmpty());
+            CHECK(string.getData() != nullptr);
+        }
+
+        SUBCASE("Wide character constructor")
+        {
+            const sf::String string(L'b');
```
[the remark this answers is not in the document]
Comment (Bromeon) [reply test/System/String.test.cpp:68]: The [2.5.1 docs ](https://en.sfml-dev.org/documentation/2.5.1/classsf_1_1String.php#ada5d5bba4528aceb0a1e298553e6c30a)of `sf::String::toAnsiString()` say:

> The UTF-32 string is converted to an ANSI string in the encoding defined by locale. **Characters that do not fit in the target encoding are discarded from the returned string.**

So they're simply removed. Do we want to test that as well?

Same with `toWideString()`, by the way.

File: test/System/String.test.cpp [hunk 4]
```diff
@@ -1,8 +1,370 @@
...
+            CHECK(string.toUtf16() == std::basic_string<std::uint16_t>{'m', 'n', 'o', 'p'});
+            CHECK(string.toUtf32() == std::basic_string<std::uint32_t>{'m', 'n', 'o', 'p'});
+            CHECK(string.getSize() == 4);
+            CHECK(!string.isEmpty());
+            CHECK(string.getData() != nullptr);
+        }
+
+        SUBCASE("UTF-32 C string constructor")
+        {
+            const sf::String string(L"qrs");
```
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/String.test.cpp]: Fixed

File: test/System/String.test.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/String.test.cpp]: It turns out that non-ASCII characters get converted into null (`\0`) characters.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
Oh joy. It appears that many of the tests fail on Windows. Is this a bug in my tests or is this simply the reality of the implementation-defined behavior of wide strings? If we're dealing with implementation-defined behavior then I'm not sure whether we can test most of these things to any meaningful degree.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
I think I found a satisfactory way to handle the fact that wide string encoding different on a per-platform basis. I'm using the width of a `wchar_t` as a proxy for which encoding its using. Then I have this function that lets us specify both potential return values for a function and the correct one will get returned.

```cpp
 // Return either argument depending on whether wchar_t is 16 or 32 bits
// Lets us write tests that work on both Windows where wchar_t is 16 bits
// and elsewhere where it is 32. Otherwise the tests would only work on
// one OS or the other.
template <typename T>
T select(const T& t16, const T& t32)
{
    assert(t16 != t32);
    if constexpr (sizeof(wchar_t) == 2)
        return t16;
    else
        return t32;
}
```

A better function name probably exists but I kept simple to begin with.

**Comment by ChrisThrasher** [comment]
I had to make `sf::String::InvalidPos` not `constexpr` to fix some MinGW linker errors. Not sure why MinGW doesn't agree with every other implementation but it's a small price to pay. Reminds me of the woes we had with the `sf::Vector` constants that we couldn't make `constexpr`.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: test/System/String.test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/String.test.cpp:314]: I addressed this by removing `constexpr`. Not ideal but it's a pragmatic choice.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
I think it's all good now

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: test/System/String.test.cpp [hunk 5]
```diff
@@ -1,8 +1,401 @@
...
+            CHECK_FALSE(sf::String() > sf::String());
+        }
+
+        SUBCASE("operator<=")
+        {
+            CHECK(sf::String() <= sf::String());
+            CHECK(sf::String('a') <= sf::String('b'));
+        }
+
+        SUBCASE("operator<")
```
Comment (kimci86) [inline test/System/String.test.cpp:389]:
```suggestion
        SUBCASE("operator>=")
```

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
SFMLBuildMaster: Build this please.

File: test/System/String.test.cpp [hunk 6]
```diff
@@ -1,8 +1,401 @@
...
+        CHECK(string.toUtf16() == std::basic_string<std::uint16_t>{'w', 'x', 'y', 'z'});
+        CHECK(string.toUtf32() == std::basic_string<std::uint32_t>{'w', 'x', 'y', 'z'});
+        CHECK(string.getSize() == 4);
+        CHECK(!string.isEmpty());
+        CHECK(string.getData() != nullptr);
+    }
+
+    SUBCASE("fromUtf32()")
+    {
+        constexpr std::array<std::uint32_t, 4> characters{'w', 'x', 'y', 'z'};
```
Comment (eXpl0it3r) [inline test/System/String.test.cpp:252]: Would it make sense to test non-ascii characters here?

File: test/System/String.test.cpp [hunk 7]
```diff
@@ -1,8 +1,401 @@
...
+        CHECK(string.getSize() == 4);
+        CHECK(!string.isEmpty());
+        CHECK(string.getData() != nullptr);
+    }
+
+    SUBCASE("clear()")
+    {
+        sf::String string("you'll never guess what happens when you call clear()");
+        string.clear();
+        CHECK(string == sf::String());
```
Comment (eXpl0it3r) [inline test/System/String.test.cpp:270]: We should assert the string size here

File: test/System/String.test.cpp [hunk 8]
```diff
@@ -1,8 +1,401 @@
...
+    }
+
+    SUBCASE("erase()")
+    {
+        sf::String string("what if i want a shorter string?");
+        string.erase(0, 8);
+        string.erase(string.getSize() - 1, 1);
+        CHECK(string == "i want a shorter string");
```
Comment (eXpl0it3r) [inline test/System/String.test.cpp:278]: We should assert the string size here

File: test/System/String.test.cpp [hunk 9]
```diff
@@ -1,8 +1,401 @@
...
+    }
+
+    SUBCASE("insert()")
+    {
+        sf::String string("please insert text");
+        string.insert(7, "don't ");
+        CHECK(string == "please don't insert text");
```
Comment (eXpl0it3r) [inline test/System/String.test.cpp:285]: We should assert the string size here

**Review by eXpl0it3r — CHANGES_REQUESTED** [review]
Some minor questions and I think we should "document" `getSize` when clearing, erasing or adding.

File: test/System/String.test.cpp [hunk 10]
```diff
@@ -1,8 +1,401 @@
...
+        {
+            if (character >= 32 && character < 127)
+                output += std::string(1, static_cast<char>(character)).c_str();
+            else
+                output += toHex(character).c_str();
+        }
+        return output;
+    }
+};
+} // namespace doctest
```
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/String.test.cpp:59]: We need to string-ify some `std::` types. Opening `namespace std` to define `toString` or `operator<<` is UB or at least it's not a good idea. The only other option we have is to specialize `doctest::StringMaker` which is an explicit part of Doctest's interface. I'm not sure what's misleading there.

File: test/System/String.test.cpp (hunk 6 above)
Comment (ChrisThrasher) [reply test/System/String.test.cpp:252]: Yeah I think I just overlooked that when I was added on-ASCII characters to the various constructors.

File: test/System/String.test.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (eXpl0it3r) [reply test/System/String.test.cpp:59]: The comment assumes that it's obvious to the reader, that one would need to open the std namespace, which isn't really given. And then the code proceeds to define something in the doctest namespace, without any explanation why it's necessary to "intrude" on a third-party namespace. The "problem" of having to stringify something, was never presented, so the comment talking about `std::` while the code implementing `doctest::` is more confusing than enlightening. 😄 

But I'll leave it up to you, whether you want to change it. 🙂

**Review by eXpl0it3r — APPROVED** [review]

File: test/System/String.test.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (ChrisThrasher) [reply test/System/String.test.cpp:59]: https://github.com/doctest/doctest/blob/master/doc/markdown/stringification.md#docteststringmakert-specialisation

Should I just use the above link? Would that be clearer?

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#sl3-do-not-add-non-standard-entities-to-namespace-std

There's also this guideline. Do we want to put all this in the comments to fully explain why we chose this particular stringification mechanism?

[the remark this answers is not in the document]
Comment (eXpl0it3r) [reply test/System/String.test.cpp:59]: The first link seems quite informative and shows that doctest encourages the opening of its own namespace for certain specialization. The later I think is quite common knowledge, the one shouldn't touch the std namespace.

Personally, I'd just write something like:

```
// Specialize StringMaker for wchar support in doctest assertions/checks
// https://github.com/doctest/doctest/blob/master/doc/markdown/stringification.md#docteststringmakert-specialisation
```

That we didn't hack us into the `std` namespace, is IMHO kind of a given and doesn't need to be mentioned.

Unless I still misunderstand the problem scope 😄

**Comment by ChrisThrasher** [comment]
I added more comments to clarify what's going on with these `doctest::StringMaker` specializations

SFMLBuildMaster: Build this please.

**Comment by ChrisThrasher** [comment]
> I think the tests may benefit from using some actual wide-char or UTF-32 characters where applicable, so we don't just have ASCII. See https://stackoverflow.com/q/6796157 for string literals 🙂

https://github.com/SFML/SFML/pull/3418

There you go :)
